### PR TITLE
Progress api

### PR DESCRIPTION
**Pull request by blagoev**
Fixes #1279

### CODE REVIEW

**Comment by blagoev** [comment]
adding for review without the docs, so we can skip updating documentation if something is changed. Use the typescript definition for quick reference.
documentation is still in progress.

**Comment by blagoev** [comment]
For reviewers,  pay attention to how Realm.open and Realm.openAsync work. Is this a good way to do it or we could support all progress modes/directions in there.

File: lib/index.d.ts [hunk 1]
```diff
@@ -411,17 +422,20 @@ declare class Realm {
...
-    static open(config: Realm.Configuration): Promise<Realm>
+    static open(config: Realm.Configuration): ProgressPromise;
     /**
      * Open a realm asynchronously with a callback. If the realm is synced, it will be fully synchronized before it is available.
      * @param {Configuration} config 
+     * @param {ProgressNotificationCallback} progressCallback? a progress notification callback for 'download' direction and 'forCurrentlyOutstandingWork' mode 
      * @param {Function} callback will be called when the realm is ready.
      */
-    static openAsync(config: Realm.Configuration, callback: (error: any, realm: Realm) => void): void
+    static openAsync(config: Realm.Configuration, progressCallback?: Realm.Sync.ProgressNotificationCallback, callback: (error: any, realm: Realm) => void): void
```
Comment (kneth) [inline lib/index.d.ts]: Would be better to have the optional argument as the last argument?

File: src/js_realm.hpp [hunk 2]
```diff
@@ -586,9 +586,14 @@ void RealmClass<T>::get_sync_session(ContextType ctx, ObjectType object, ReturnV
 
 template<typename T>
 void RealmClass<T>::wait_for_download_completion(ContextType ctx, FunctionType, ObjectType this_object, size_t argc, const ValueType arguments[], ReturnValue &return_value) {
-    validate_argument_count(argc, 2);
+    validate_argument_count(argc, 2, 3);
     auto config_object = Value::validated_to_object(ctx, arguments[0]);
-    auto callback_function = Value::validated_to_function(ctx, arguments[1]);
+    auto callback_function = Value::validated_to_function(ctx, arguments[argc - 1]);
```
Comment (kneth) [inline src/js_realm.hpp]: Would be better to have the optional argument as the last argument?

File: src/js_sync.hpp [hunk 3]
```diff
@@ -170,6 +171,10 @@ class SessionClass : public ClassDefinition<T, WeakSession> {
 
     static void simulate_error(ContextType, FunctionType, ObjectType, size_t, const ValueType[], ReturnValue &);
     static void refresh_access_token(ContextType, FunctionType, ObjectType, size_t, const ValueType[], ReturnValue &);
+#if REALM_ENABLE_SYNC
```
Comment (kneth) [inline src/js_sync.hpp]: `src/js_sync.hpp` is only included if `REALM_ENABLE_SYNC` is defined so I think we don't need this.

File: tests/js/download-api-helper.js [hunk 4]
```diff
@@ -2,33 +2,49 @@
...
+}
 
+Realm.Sync.User.register('http://localhost:9080', username, 'password', (error, registeredUser) => {
+    if (error) {
+        const registrationError = JSON.stringify(error);
+        Realm.Sync.User.login('http://localhost:9080', username, 'password', (err, loggedUser) => {
+            if (err) {
+                const loginError = JSON.stringify(err);
+                console.error("download-api-helper failed:\n User.register() error:\n" + registrationError + "\n User.login() error:\n" + loginError);
+                process.exit(-2);
```
Comment (kneth) [inline tests/js/download-api-helper.js]: Maybe throw an exception. Termination does look a bit too rough to me.

Comment (blagoev) [reply tests/js/download-api-helper.js]: this file is run out of process so it's on purpose

File: lib/index.d.ts (hunk 1 above)
Comment (blagoev) [reply lib/index.d.ts]: well the callback is the last as per nodejs convention. So this is stuffed before that. Also the js code allows to call the function with openAsync(config, callback) as before. Maybe I could change this to a default initialized parameter

File: src/js_sync.hpp (hunk 3 above)
Comment (blagoev) [reply src/js_sync.hpp]: will remove

File: lib/index.d.ts (hunk 1 above)
Comment (kneth) [reply lib/index.d.ts]: My question has clearly non-JS roots :smile: Let's leave it as it is.

**Comment by blagoev** [comment]
docs https://github.com/realm/realm.io/pull/3793

File: src/js_sync.hpp [hunk 5]
```diff
@@ -308,7 +313,83 @@ void SessionClass<T>::refresh_access_token(ContextType ctx, FunctionType, Object
...
+        Protected<typename T::GlobalContext> protected_ctx(Context<T>::get_global_context(ctx));
+        std::function<ProgressHandler> progressFunc; 
+
+        EventLoopDispatcher<ProgressHandler> progress_handler([=](uint64_t transferred_bytes, uint64_t transferrable_bytes) {
+            HANDLESCOPE
+            ValueType callback_arguments[2];
+            callback_arguments[0] = Value::from_number(protected_ctx, transferred_bytes);
+            callback_arguments[1] = Value::from_number(protected_ctx, transferrable_bytes);
+
+            Function<T>::callback(protected_ctx, protected_callback, protected_this, 2, callback_arguments);
```
Comment (fealebenpae) [inline src/js_sync.hpp]: This way the `this` of the callback is the session, but if the callback is an arrow function that's sort of useless. Wouldn't it be better to pass it as an argument instead?

File: src/js_sync.hpp [hunk 6]
```diff
@@ -308,7 +313,83 @@ void SessionClass<T>::refresh_access_token(ContextType ctx, FunctionType, Object
...
+        });
+
+        progressFunc = std::move(progress_handler);
+
+        
+        auto registrationToken = session->register_progress_notifier(std::move(progressFunc), notifierType, false);
+
+        auto syncSession = create_object<T, SessionClass<T>>(ctx, new WeakSession(session));
+        PropertyAttributes attributes = ReadOnly | DontEnum | DontDelete;
+        Object::set_property(ctx, callback_function, "syncSession", syncSession, attributes);
```
Comment (fealebenpae) [inline src/js_sync.hpp]: I know the properties are non-enumerable (though I think they would still show in a debugger), but I suggest prepending the names with an underscore. I'm not sure what your thoughts on the matter are, but I treat underscored member names in JavaScript as a convention to demonstrate that something is private.

File: src/js_sync.hpp (hunk 5 above)
Comment (blagoev) [reply src/js_sync.hpp]: which `this` should be better used here for the callback when called. (from user perspective)?

File: src/js_sync.hpp (hunk 6 above)
Comment (blagoev) [reply src/js_sync.hpp]: sure will fix

File: src/js_sync.hpp (hunk 5 above)
Comment (fealebenpae) [reply src/js_sync.hpp]: I think not specifying any `this` at this level would be best.

Comment (blagoev) [reply src/js_sync.hpp]: 👍

**Comment by kneth** [comment]
The failing test is:

```
PermissionTests : testApplyAndGetGrantedPermissions
   TypeError: Cannot read property 'path' of undefined
       at user.applyPermissions.then.then.permissions (/Users/realm/workspace/realm_realm-js_PR-1279-V5WM5CA2KFXWHKK7MY3PF3THQSTPR7NRC3MKNYVZK7FID7SDH6KQ/tests/js/permission-tests.js:64:50)
```

**Comment by blagoev** [comment]
yes indeed. the reason is permission tests are failing sporadically. will do a rebuild. we will need to look at that if this happens too much.

**Pull request merged**